Feature: Add `eq_with_nullability_subset ` and `eq_with_nullability_superset ` and validate `extend_from_array`

connortsui20 opened the pull request:
This will be helpful in my refactoring of the `builders` module in `vortex-array` (in draft right now: https://github.com/vortex-data/vortex/pull/4440)

Note that `extend_from_array` is called in many non-test places yet is not validated at all. Given that this is a public method, this is a bug waiting to happen...

I have an interesting thought though: should a `DType::Primitive(U64, nullable)` be a superset of `DType::Primitive(U16, non-nullable)`?

In theory, we should be able to extend an array builder of nullable `U64` with an array of non-nullable `U8`.

I'll leave that to a separate PR since this idea might have some other implications.

connortsui20 committed:
add `eq_with_nullability_subset` and `eq_with_nullability_superset`

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
validate `extend_from_array` in all builders

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
clean up some builders

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
Chore: Remove `append_scalar_value` (#4447)

The method doesn't really make sense, especially since `ScalarValue`
doesn't actually hold nullability.

Also refactors the `random_array` function.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

robert3005 reviewed (changes requested):
Instead of introducing type hierarchy there's maybe a simpler thing to do here where you check for nullability. I don't see how a builder can optimize the append any differently than the caller. This is worth revisiting imho when we add a downcast array.

connortsui20 commented:
That makes sense, I'll make an issue for it. However, this PR doesn't actually introduce a type hierarchy and considering we don't actually use type hierarchy anywhere else I feel like this is fine to merge as is. Or do you think the name needs to change?

connortsui20 committed:
Chore: Add constant default builder capacity (#4448)

Adds a constant default builder capacity instead of repeating 1024
everywhere.

---------

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
Chore: Clean up `DecimalBuilder` file and implementation (#4449)

Mostly just shuffles stuff around to make things more readable, but I
also removed several methods that were somewhat redundant.

---------

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

robert3005 commented:
I would change the name since people could assume that it does more than what it says. Otherwise I think this change is gtg

connortsui20 commented:
Renamed to `eq_with_nullability_subset` and `eq_with_nullability_superset`

connortsui20 commented:
Also I just realized I messed up the branch chain, none of the other PRs I made last night got merged to develop, they got merged into this branch
